src/node_task_queue.cc

```diff
@@ -76,12 +113,46 @@ void PromiseRejectCallback(PromiseRejectMessage message) {
 
   Local<Value> args[] = { type, promise, value };
 
-  // V8 does not expect this callback to have a scheduled exceptions once it
-  // returns, so we print them out in a best effort to do something about it
-  // without failing silently and without crashing the process.
+  double async_id = AsyncWrap::kInvalidAsyncId;
+  double trigger_async_id = AsyncWrap::kInvalidAsyncId;
   TryCatchScope try_catch(env);
+
+  if (!GetAssignedPromiseAsyncId(env, promise, env->async_id_symbol())
+          .To(&async_id)) return;
+  if (!GetAssignedPromiseAsyncId(env, promise, env->trigger_async_id_symbol())
+          .To(&trigger_async_id)) return;
+
+  if (async_id == AsyncWrap::kInvalidAsyncId &&
+      trigger_async_id == AsyncWrap::kInvalidAsyncId) {
+    // That means that promise might be a PromiseWrap, so we'll
+    // check there as well.
```

Comment:
Can we add tests for both cases here?

Reply:
I actually manually ran the other case by commenting [this](https://github.com/nodejs/node/blob/master/test/async-hooks/init-hooks.js#L51) line out while running the test case added in this test. It passed. 

However, I guess we can change `ActivityCollector` to not use a `noop` function when `ondestroy` is not provided. What do you think?

Reply:
I'm not sure why the test would need to use `ActivityCollector`, or be in `test/async-hooks` -- a lot of async_hooks tests that don't make use of the special features that that subdirectory has are just in `test/parallel`.

Reply:
I see. What are these "special features" out of curiosity?

Also, if I just assumed initHooks had convenient methods when dealing with async hooks. What is otherwise the purpose?

Reply:
> What are these "special features" out of curiosity?

Things like `ActivityCollector`, and the rest that’s in `async-hooks/init-hooks.js` and `async-hooks/hook-checks.js`.

> Also, if I just assumed initHooks had convenient methods when dealing with async hooks. What is otherwise the purpose?

They do, but it’s not like you *have* to use them. At least for the test here, the difference to using async hooks directly seems to be relatively small.